numba/_dispatcher.c

```diff
@@ -191,18 +268,53 @@ int search_new_conversions(PyObject *dispatcher, PyObject *args, PyObject *kws)
 
 /* A custom, fast, inlinable version of PyCFunction_Call() */
 static PyObject *
-call_cfunc(PyObject *cfunc, PyObject *args, PyObject *kws)
+call_cfunc(DispatcherObject *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyObject *locals)
 {
     PyCFunctionWithKeywords fn;
+#if PROFILE_CALL
+    PyThreadState *tstate;
+#endif
     assert(PyCFunction_Check(cfunc));
     assert(PyCFunction_GET_FLAGS(cfunc) == METH_VARARGS | METH_KEYWORDS);
     fn = (PyCFunctionWithKeywords) PyCFunction_GET_FUNCTION(cfunc);
-    return fn(PyCFunction_GET_SELF(cfunc), args, kws);
+#if PROFILE_CALL
+    tstate = PyThreadState_GET();
+    if (tstate->use_tracing && tstate->c_profilefunc)
+    {
+        /*
+	 * The following code requires some explaining:
+	 * The PyFrame_New() constructor doesn't do anything with the 'locals' value if the 'code's
+	 * 'CO_NEWLOCALS' flag is set (which is always the case nowadays).
+	 * So, to get local variables into the frame, we have to manually set the 'f_locals'
+	 * member, then call `PyFrame_LocalsToFast`, where a subsequent call to the `frame.f_locals`
+	 * property (by virtue of the `frame_getlocals` function in frameobject.c) will find them.
+	 */
+        PyCodeObject *code = (PyCodeObject*)PyObject_GetAttrString((PyObject*)self, "func_code");
+        PyObject *globals = PyDict_New();
+        PyFrameObject *frame = PyFrame_New(tstate, code, globals, NULL);
+        // now populate the 'fast locals' in `frame`
+        frame->f_locals = locals;
+        PyFrame_LocalsToFast(frame, 0);
+        frame->f_locals = NULL;
```

Comment:
The original locals is a `borrowed reference` (from a call to PyEval_GetLocals()), so I didn't think it had to be decrefed in the first place. Does PyFrame_LocalsToFast() increment the reference to it ?